feat: iOS and Android native support

munkiki7 opened the pull request:
### Description

Add full support for iOS crash reporting in Unity games(bugsplat-unity #35):
- Add an option to toggle functionality on/off in Unity Editor
- Add [bugsplat-ios](https://github.com/bugsplat-git/bugsplat-ios) to generated iOS project
- Configure automatic symbol uploads

Provide the scaffolding around Crashpad for Android devices(bugsplat-unity #36):
- Add an option to toggle functionality on/off in Unity Editor
- Link the Crashpad libraries
- Include the Crashpad handler executable in the build

Resolves #35 
Resolves #36

bobbyg603 commented:
Hi @munkiki7, thank you for this PR - there's some really excellent work here! Apologies for the delay on the review.

I'm still reviewing, but I think I found at least 1 item that could use some re-work while you're waiting for me to finish the review. It looks like the iOS and Android reporting was rework to use either the .NET handler or the Native handler instead of both for their appropriate use cases. I think what we really want is the .NET Exception Handler enabled at all times, and the option to additionally add Native reporting. I believe this is how we handle Windows for instance.

Does this make sense? Please let me know if you have any concerns or need clarification.

bobbyg603 commented:
@munkiki7 can you please add `I Phone` and `Android` platforms to the native crash buttons in the sample? I have a commit for this, but can't seem to update this PR. It's probably easiest if you attempt to recreate this work.

Steps:
1. Create a new Unity project and add the `bugsplat-unity` package via a path to your repo on the local file system
2. In Package Manager window, import the bugsplat-unity sample
3. Add iPhone and Android to the list of platforms for the following buttons `Button_AccessViolation`, `Button_FatalError`, `Button_Abort`, `Button_PureVirtualFunction`, and `Button_MonoAbort`. Follow the example below

https://user-images.githubusercontent.com/2646053/202236129-b0dc558c-bcfe-4aac-b4f4-249a6391996a.mov

4. Copy the scene back into the git repo and commit it as part of this PR

munkiki7 commented:
@bobbyg603 , sorry for the delay. I have changed the demo scene and have changed the logic for the dotnet handler on android and iOS. I will look into the version match a bit later.

munkiki7 commented:
@bobbyg603 , today we have looked into the version issue for the iOS symbols. You were correct. Once we have changed the version to represent the actual app version string (1.0 (3) in our case), the reports were symbolicated. We have also added a native iOS crash method to demo.

Here are the links to [native](https://app.bugsplat.com/v2/crash?database=unity_plugin_test&id=37) and [Unity](https://app.bugsplat.com/v2/crash?database=unity_plugin_test&id=36) crash reports.

bobbyg603 commented:
Hi @munkiki7, thanks for your patience. I still appear to be running into a version mismatch here

> Hi @munkiki7, thank you for this PR - there's some really excellent work here! Apologies for the delay on the review. I'm still reviewing, but I think I found at least 1 item that could use some re-work while you're waiting for me to finish the review.
> 
> It looks like the iOS and Android reporting was changed to use either the .NET handler or the Native handler instead of both for their appropriate use cases. I think what we really want is the .NET Exception Handler enabled at all times, and the option to additionally add Native reporting. We have a similar implementation on Windows where the .NET handler is enabled at all times, and the Unity native handler catches native crashes.
> 
> Does this make sense? Please let me know if you have any concerns or need clarification.

This has been fixed, nice work!

bobbyg603 commented:
> @bobbyg603 , sorry for the delay. I have changed the demo scene and have changed the logic for the dotnet handler on android and iOS. I will look into the version match a bit later.

I've verified that this has been fixed - thank you!

bobbyg603 commented on Editor/PostBuild.cs:
```diff
@@ -1,130 +1,229 @@
+		}
+
+		var projectDir = Path.GetDirectoryName(Application.dataPath);
+		if (projectDir == null)
```
```suggestion
		if (string.IsNullOrEmpty(projectDir))
```

bobbyg603 commented on Editor/PostBuild.cs:
```diff
@@ -1,130 +1,229 @@
-    {
-        var options = GetBugSplatOptions();
-
-        if (options == null)
```
Why did this get removed?

munkiki7 replied:
The null check is now present in the top-most method.
`public static async Task OnPostprocessBuild(BuildTarget target, string pathToBuiltProject)`

bobbyg603 replied:
👍 must've missed that, thanks!

bobbyg603 commented on Editor/PostBuild.cs:
```diff
@@ -1,130 +1,229 @@
+
+		const string bugSplatServerURLKey = "BugsplatServerURL";
+		plistInfoFile.root.AsDict().SetString(bugSplatServerURLKey, $"https://{options.Database}.bugsplat.com/");
+
```
I could use a bit of help understanding our options for version here. I think perhaps this is our hook into getting the versions matching for our symbol uploads. 

* In Unreal I believe we chopped of the build revision (the (3) part in 1.0 (3)).
    * Can we make a similar change in Unity so that there's only one source of truth for the application's version? 
    * If the user builds the native solution again, will this step be re-run and update the version in Info.plist?
* Or should we instruct the user to manage their version number manually?

Any details you can add are greatly appreciated!

munkiki7 replied:
Yes, the best solution would be to use the application version, as it is already managed by the user in the project settings.

It would be redundant to make the user type the app version in two different places.

bobbyg603 commented on Samples~/my-unity-crasher/BugSplatOptions.asset:
```diff
@@ -12,9 +12,9 @@ MonoBehaviour:
-  Database: fred
-  Application: 
-  Version: 
+  Database: unity_plugin_test
+  Application: BugSplatDemo
+  Version: 1.0 (3)
```
```suggestion
  Database: fred
  Application: 
  Version: 
```

bobbyg603 replied:
This is related to the [comment](https://github.com/BugSplat-Git/bugsplat-unity/pull/74/files#r1035091838) above.

bobbyg603 commented on Samples~/my-unity-crasher/Scripts/Crasher.cs:
```diff
@@ -27,6 +27,13 @@ public void Event_ForceCrash(ForcedCrashCategory category)
 			Utils.ForceCrash(category);
 		}
+		
+		public void Event_NativeCrashIos()
```
Now that the native crash buttons are enabled in the sample do we need this?

munkiki7 replied:
Those methods are still C# crashes, while this one is native objc crash. 

It is good to have it for checking of ObjectiveC crashes.

bobbyg603 replied:
👍
